### DISCUSSION TITLE
Using Box<dyn Trait> with extensions

### DISCUSSION THREAD

**Discussion by Datron** [actix/actix-web#3335] [Q&A]
Hi folks,

I have a trait - lets say 
```
pub trait UserTrait {
    fn get_email(&self) -> String;
    fn get_username(&self) -> String;
    fn get_auth_token(&self) -> String;
}
```
a struct like this:
```
pub struct UserWrapper(pub Box<dyn UserTrait>);
```
and a struct that implements this trait - 

```
struct User;

impl UserTrait for User {
// do something
}
```

Now, in some middleware I want to create struct User, set it in extensions like so:
```
let user = get_user_somehow();
req.extensions_mut().insert::<UserWrapper>(user);
```

So that I can read this in the handler and call the trait functions on it like so:
```
#[put("/{key}")]
async fn create(
    state: Data<AppState>,
    key: web::Path<String>,
    raw_request: HttpRequest,
    request: web::Json<CreateReq>,
    db_conn: DbConnection,
) -> Result<HttpResponse> {
    let extensions = raw_request.extensions();
    let user: &UserWrapper = extensions.get::<UserWrapper>().unwrap();
    // rest of code
}
```
When I use the debugger I can see the type stored in extensions, but it is not being retrieved and my code panics when I unwrap(). What am I doing wrong here? 

I'm sorry if this has been asked before, I didn't find anything on searching. Thank you for the library!

**Comment by robjtede** [comment]
If the middleware is always setting the request extension and is wrapping everything (at the App level) then this should work. If you're still having issues and think there's a bug then please open an issue with a complete minimal reproduction.
